src/Shared/Microsoft.AspNetCore.Razor.Test.Common/IntegrationTestAttributes.cs

```diff
@@ -0,0 +1,90 @@
+[XunitTestCaseDiscoverer("Microsoft.AspNetCore.Razor.Test.Common." + nameof(IntegrationTheoryDiscoverer), "Microsoft.AspNetCore.Razor.Test.Common")]
+internal class IntegrationTestTheoryAttribute : TheoryAttribute { }
+
+internal class IntegrationFactDiscoverer(IMessageSink diagnosticMessageSink)
+    : FactDiscoverer(diagnosticMessageSink)
+{
+    public override IEnumerable<IXunitTestCase> Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
+    {
+        return [
+            new IntegrationTestCase(designTime: true, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod),
+            new IntegrationTestCase(designTime: false, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod)
+        ];
+    }
+}
+
+internal class IntegrationTheoryDiscoverer(IMessageSink diagnosticMessageSink)
+    : TheoryDiscoverer(diagnosticMessageSink)
+{
+    protected override IEnumerable<IXunitTestCase> CreateTestCasesForDataRow(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute, object[] dataRow)
+    {
+        return [
+            new IntegrationTestCase(designTime: true, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod, dataRow),
+            new IntegrationTestCase(designTime: false, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod, dataRow)
+        ];
+    }
+}
+
+internal class IntegrationTestCase : XunitTestCase
+{
+    private bool designTime;
+
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes")]
+    public IntegrationTestCase() { }
+
+    public IntegrationTestCase(bool designTime, IMessageSink diagnosticMessageSink, TestMethodDisplay defaultMethodDisplay, TestMethodDisplayOptions defaultMethodDisplayOptions, ITestMethod testMethod, object[] testMethodArguments = null!)
+        : base(diagnosticMessageSink, defaultMethodDisplay, defaultMethodDisplayOptions, testMethod, testMethodArguments)
+    {
+        this.designTime = designTime;
+    }
+
+    protected override string GetDisplayName(IAttributeInfo factAttribute, string displayName)
+    {
+        return base.GetDisplayName(factAttribute, displayName) + (designTime ? " (Design Time)" : " (Run Time)");
+    }
+
+    public override Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink, IMessageBus messageBus, object[] constructorArguments, ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource)
+    {
+        Debug.Assert(constructorArguments.Length == 1 && constructorArguments[0] is bool);
+        constructorArguments[0] = designTime;
```

Comment:
I did not realize you could use a custom xunit test case to pass in constructor arguments. That is really nice. Wish I'd realized that when I was doing some of the `WpfFact` work.